File: java/src/Horse.java
import java.util.*;

public class Horse {

    public static void main(String[] args) {
        int attempts = Integer.parseInt(args[0]);
        int size_x = Integer.parseInt(args[1]);
        int size_y = Integer.parseInt(args[2]);

        Date time_0 = new Date();
        for (int attempt = 0; attempt < attempts; attempt++) {
            calc_horse(size_x, size_y, 0, 0);
            System.out.println();
        }
        Date time_1 = new Date();

        double duration = (time_1.getTime() - time_0.getTime()) / 1000.0;
        System.out.println("---");
        System.out.println(String.format("Attempts: %s, duration: %s sec", attempts, duration));
    }

    /**
     * Store for statistic
     */
    static class State {
        int path_count_total = 0;
        int path_count_ok = 0;
    }

    /**
     * Calc and print all full knight's tours over specified board
     *
     * @param size_x board size x
     * @param size_y board size y
     * @param x0     tour start cell x
     * @param y0     tour start cell y
     */
    static void calc_horse(int size_x, int size_y, int x0, int y0) {
        System.out.println(String.format("Hello, horse, board %sx%s", size_x, +size_y));

        int[][] grid = new int[size_y][size_x];

        State state = new State();

        Date time_0 = new Date();
        step_horce(1, state, x0, y0, grid);
        Date time_1 = new Date();

        System.out.println(String.format("Board %sx%s, full path count: %s, total path count: %s", size_x, size_y, state.path_count_ok, state.path_count_total));

        double duration_sec = (time_1.getTime() - time_0.getTime()) / 1000.0;
        System.out.println("duration: " + duration_sec + " sec");
    }

    private static void step_horce(int step_no, State state, int x0, int y0, int[][] position) {
        // ---------------------
        // Make my step
        position[y0][x0] = step_no;

        // ---------------------
        // Print board after my step
        // print_board(position);
        // System.out.println();

        // ---------------------
        // Try to do next 8 steps
        int size_y = position.length;
        int size_x = position[0].length;
        int board_size = size_x * size_y;

        int[][] step_diffs = {{1, -2}, {2, -1}, {2, 1}, {1, 2}, {-1, 2}, {-2, 1}, {-2, -1}, {-1, -2}};

        int steps_done = 0;
        for (int s = 0; s < 8; s++) {
            int[] step_diff = step_diffs[s];
            int x1 = x0 + step_diff[0];
            int y1 = y0 + step_diff[1];

            if (x1 < 0 || x1 >= size_x) {
                continue;
            }
            if (y1 < 0 || y1 >= size_y) {
                continue;
            }
            if (position[y1][x1] != 0) {
                continue;
            }

            step_horce(step_no + 1, state, x1, y1, position);

            steps_done = steps_done + 1;
        }

        // ---------------------
        // Whe have no more cell to step?
        if (steps_done == 0) {
            state.path_count_total = state.path_count_total + 1;
        }

        if (steps_done == 0 && step_no == board_size) {
            state.path_count_ok = state.path_count_ok + 1;
            // System.out.println(String.format("Full path count: %s/%s", state.path_count_ok, state.path_count_total));
            // print_board(position);
            // System.out.println();
        }

        // ---------------------
        // Make my step back
        position[y0][x0] = 0;
    }

    private static void print_board(int[][] position) {
        int size_y = position.length;
        int size_x = position[0].length;

        int y = 0;
        while (y < size_y) {
            int x = 0;
            while (x < size_x) {
                int v = position[y][x];
                if (v == 0) {
                    System.out.print(String.format("%3s", "*") + " ");
                } else {
                    System.out.print(String.format("%3s", v) + " ");
                }
                x = x + 1;
            }
            System.out.println();
            y = y + 1;
        }
    }

}

File: python/horse.py
import datetime
import sys


def print_board(position):
    size_y = len(position)
    size_x = len(position[0])

    y = 0
    while (y < size_y):
        x = 0
        while (x < size_x):
            v = position[y][x]
            if (v == 0):
                print("*".rjust(3) + " ", end="")
            else:
                print(str(v).rjust(3) + " ", end="")
            x = x + 1

        print()
        y = y + 1


def step_horce(step_no, state, x0, y0, position):
    step_diffs = [[1, -2], [2, -1], [2, 1], [1, 2], [-1, 2], [-2, 1], [-2, -1], [-1, -2]]

    # ---------------------
    # Make my step
    position[y0][x0] = step_no

    # ---------------------
    # Print board after my step
    # print_board(position)
    # print()

    # ---------------------
    # Try to: next 8 step
    size_y = len(position)
    size_x = len(position[0])
    board_size = size_x * size_y

    steps_done = 0
    for s in range(8):
        x1 = x0 + step_diffs[s][0]
        y1 = y0 + step_diffs[s][1]

        if (x1 < 0) or (x1 >= size_x):
            continue

        if (y1 < 0) or (y1 >= size_y):
            continue

        if (position[y1][x1] != 0):
            continue

        step_horce(step_no + 1, state, x1, y1, position)

        steps_done = steps_done + 1

    # ---------------------
    # Whe nave no more steps?
    if (steps_done == 0):
        state["path_count_total"] = state["path_count_total"] + 1

    if (steps_done == 0) and (step_no == board_size):
        state["path_count_ok"] = state["path_count_ok"] + 1
        # print("Full path count: " + str(state["path_count_ok"]) + "/" + str(state["path_count_total"]))
        # print_board(position)
        # print("")

    # ---------------------
    # Make my step back
    position[y0][x0] = 0


def diff_msec(dt_2, dt_1):
    done_msec = (dt_2 - dt_1).days * 24 * 60 * 60 * 1000 + (dt_2 - dt_1).seconds * 1000 + (dt_2 - dt_1).microseconds / 1000
    return done_msec


def calc_horse(size_x, size_y, x0, y0):
    print("Hello, horse, board " + str(size_x) + "x" + str(size_y))

    grid = [[0] * size_x for i in range(size_y)]

    state = {"path_count_ok": 0, "path_count_total": 0}

    time_0 = datetime.datetime.now()
    step_horce(1, state, x0, y0, grid)
    time_1 = datetime.datetime.now()

    print("Board " + str(size_x) + "x" + str(size_y) + ", full path count: " + str(state["path_count_ok"]) + ", total path count: " + str(state["path_count_total"]))

    duration_sec = int(diff_msec(time_1, time_0)) / 1000
    print("duration: " + str(duration_sec) + " sec")


def main():
    attempts = int(sys.argv[1])
    size_x = int(sys.argv[2])
    size_y = int(sys.argv[3])

    time_0 = datetime.datetime.now()
    for attempt in range(attempts):
        calc_horse(size_x, size_y, 0, 0)
        print()
    time_1 = datetime.datetime.now()

    duration = int(diff_msec(time_1, time_0)) / 1000
    print("---")
    print("Attempts: " + str(attempts) + ", duration: " + str(duration) + " sec")


###
main()
